networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/AttestationSubnetSubscriptions.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.eth2.gossip;
+
+import static java.lang.StrictMath.toIntExact;
+import static tech.pegasys.artemis.util.config.Constants.ATTESTATION_SUBNET_COUNT;
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.primitives.UnsignedLong;
+import io.vertx.core.impl.ConcurrentHashSet;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import tech.pegasys.artemis.networking.eth2.gossip.topics.AttestationTopicHandler;
+import tech.pegasys.artemis.networking.p2p.gossip.GossipNetwork;
+import tech.pegasys.artemis.networking.p2p.gossip.TopicChannel;
+import tech.pegasys.artemis.storage.client.RecentChainData;
+
+public class AttestationSubnetSubscriptions implements AutoCloseable {
+  private final GossipNetwork gossipNetwork;
+  private final RecentChainData recentChainData;
+  private final EventBus eventBus;
+
+  private final Map<Integer, Set<UnsignedLong>> subnetIdToCommittees = new ConcurrentHashMap<>();
+  private final Map<Integer, TopicChannel> subnetIdToTopicChannel = new ConcurrentHashMap<>();
+
+  public AttestationSubnetSubscriptions(
+      final GossipNetwork gossipNetwork,
+      final RecentChainData recentChainData,
+      final EventBus eventBus) {
+    this.gossipNetwork = gossipNetwork;
+    this.recentChainData = recentChainData;
+    this.eventBus = eventBus;
+  }
+
+  public Optional<TopicChannel> getChannel(final UnsignedLong committeeIndex) {
+    final int subnetId = committeeIndexToSubnetId(committeeIndex);
+    if (!subnetIdToCommittees
+        .computeIfAbsent(subnetId, __ -> Collections.emptySet())
```

Comment:
```suggestion
        .getOrDefault(subnetId, Collections.emptySet())
```

While it looks like `Collections.emptySet()` would create a new object on every call, it actually doesn't and always returns the same instance. One of the advantages of Java generics being stripped at runtime. :)

And `computeIfAbsent` would add the immutable empty set to the collection, likely causing issues later.  Suspect you should add a test for getting a channel when nothing subscribes and then later trying to subscribe to it to reveal the bug. :)

Reply:
I'm also a little unclear on why we'd want to check if we're subscribed to the specific committee index.  If we needed to publish something to a committee we aren't subscribed to but happen to be subscribed to a different committee on the same subnet, wouldn't we want to just use the subnet we are subscribed to?  Effectively when we subscribe to a subnet we subscribe to all the committee that use that subnet.

Reply:
I agree. If we are subscribed to the subnet already, there is no reason not to use it. Will make the change.

Reply:
Removed the part of code where we check for committee indices in `getChannel`. Now we only compute the subnet index and return the topic channel of that subnet index if we're subscribed to it, or an empty Optional if not.